pkg/metrics/clustermesh.go

```diff
@@ -0,0 +1,238 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2022 Authors of Cilium
+
+package metrics
+
+import (
+	"regexp"
+	"sync/atomic"
+	"time"
+
+	clientPkg "github.com/cilium/cilium/pkg/client"
+
+	"github.com/prometheus/client_golang/prometheus"
+	log "github.com/sirupsen/logrus"
+)
+
+const (
+	subsystemClusterMesh = "clustermesh"
+)
+
+var _ prometheus.Collector = (*clustermeshCollector)(nil)
+
+var (
+	// sample remote cluster etcd status
+	// etcd: 1/1 connected, lease-ID=7c027e1caebe4533, lock lease-ID=7c027e1caebe4535, has-quorum=true: https://kind-cilium-mesh-1.mesh.cilium.io:31484 - 3.4.13 (Leader)
+	clusterMeshRemoteClusterEtcdStatusRegex = regexp.MustCompile(`^etcd: \d+/\d+ connected, lease-ID=([0-9a-f]+), lock lease-ID=([0-9a-f]+), has-quorum=true`)
```

Comment:
If some portion of this code is derived from a similar implementation elsewhere such as https://github.com/form3tech-oss/extra-cilium-metrics/blob/master/main.go, I would suggest locating the original authors via `git blame` and attributing them also in the commit messages for any commits that include their code, for instance by adding `Co-Authored-By: John Doe <john@doe.org>`.

Reply:
I personally don't prefer relying on string text to grep details, as this one can be changed anytime. If we need to provide such metrics, we can expose details in more structured way.

https://github.com/cilium/cilium/blob/a8e3fa283cf40b330035265bf83d63800b296bcf/pkg/kvstore/etcd.go#L1153

Reply:
I personally don't like grepping details from free-style string, it's error prone, also others might update orignal status message to something else later. I am thinking about to expose details in more structured way in related code below. This is kind of aligned with alternative approach of exposing metrics _in line_.

https://github.com/cilium/cilium/blob/a8e3fa283cf40b330035265bf83d63800b296bcf/pkg/kvstore/etcd.go#L1153

Reply:
:+1: yeah agree, seeing a random regex somewhere in the code always raises some questions ;-)

Reply:
Based on https://github.com/cilium/cilium/pull/18313, there will be changes to remove consul and kvstore abstraction.

Currently, etcd related metrics are also abstracted, however, a couple of new metrics are kind of etcd specific, I reckon we might need to wait a little bit for the above PR to land in first.

Reply:
I think we should try to avoid messing with the metrics as part of that work; if some metrics are already abstracted, my vote would be to keep them the same. If new metrics really are specific to etcd, I'm OK with adding them with an etcd-specific name. Or if we think that they should apply for any kvstore, we could abstract the name to say 'kvstore' rather than 'etcd'.